### PR TITLE
add fenv.h

### LINKED ISSUE #2590: add cfenv binding
Some math functions depend on fenv. (e.g. rounding functions `fesetround`)

### PR DESCRIPTION
**Pull request by i10416**
close #2590 

rel #104

### CODE REVIEW

**Comment by i10416** [comment]
Sorry, ready for review now.

**Comment by i10416** [comment]
> To use this function, you must turn off floating-point optimizations that could prevent access by using the #pragma fenv_access(on) directive prior to the call.

**Comment by i10416** [comment]
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fetestexcept1?view=msvc-170

**Comment by i10416** [comment]
🤔

**Comment by i10416** [comment]
How to use `#pragma` in scala native...?

**Comment by WojciechMazur** [comment]
> 


> How to use `#pragma` in scala native...?

It's not possible in Scala sources compiled with Scala Native. You can only use it  C/C++ sources

**Comment by i10416** [comment]
> It's not possible in Scala sources compiled with Scala Native. 

Oh... the linked Microsoft documentation says that `fenv_access` is __off by default__ and developers need to explicitly set `#pragma fenv_access(on)` to avoid unintended optimization gotcha.

> What directive you wanted to use

I am not going to use any `#pragma` for now, but it is likely that someone who tries to call functions in `fenv.h` on Windows get confused, especially when they can use these functions but functions return unexpected response.

> Maybe there is a workaround for that

Yeah, if there is, I'm sure we need to mention that in scaladoc.

**Comment by i10416** [comment]
It seems `feraiseexcept` is not always supported...😖

**Comment by i10416** [comment]
Hmmm, some platform/arch add FE_INEXACT flag when FE_OVERFLOW is set.

**Comment by i10416** [comment]
I hope `if (fenv.FE_ALL_EXCEPT == 0) {` can be evaluated at compile time and compiler removes clause containing `feraiseexcept` ...

**Comment by WojciechMazur** [comment]
If it happens only in Windows (CI suggests that), we can replace this test with `if(scalanative.meta.LinktimeInfo.isWindows)` in that case the untaken branch would be dead-code eliminated when linking the NIR, and though we would get rid of undefined symbols reported by LLVM linker. It would then use a mechanism of linktime-values which were used to handle Windows support

**Comment by i10416** [comment]
Perhaps fenv tests affect other test results?

File: clib/src/main/scala/scala/scalanative/libc/fenv.scala [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+   *  @return
+   *    ​0​ on success, non-zero otherwise.
+   */
+  def feupdateenv(envp: Ptr[fenv_t]): CInt = extern
+
+  // floating-point exceptions
+
+  /** pole error occurred in an earlier floating-point operation
+   */
+  @name("scalanative_fenv_fe_divbyzero")
```
[the remark this answers is not in the document]
Comment (i10416) [reply clib/src/main/scala/scala/scalanative/libc/fenv.scala]: Removed fenv.

**Review by ekrich — CHANGES_REQUESTED** [review]

File: clib/src/main/resources/scala-native/fenv.c [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#include <fenv.h>
+
+int scalanative_fenv_fe_divbyzero() { return FE_DIVBYZERO; }
```
[the remark this answers is not in the document]
Comment (i10416) [reply clib/src/main/resources/scala-native/fenv.c]: Probably you see older version?

https://github.com/scala-native/scala-native/pull/2591/files#diff-9e88de73f7e02009f1cb7df744326b7d32b1842ff706b2efc505f6471c9ed9bd

[the remark this answers is not in the document]
Comment (i10416) [reply clib/src/main/resources/scala-native/fenv.c]: https://github.com/scala-native/scala-native/pull/2591/commits/b8de2eceecb47c29373e9cb940c61f8e71080443

[the remark this answers is not in the document]
Comment (ekrich) [reply clib/src/main/resources/scala-native/fenv.c]: Oh sorry about that.

[the remark this answers is not in the document]
Comment (i10416) [reply clib/src/main/resources/scala-native/fenv.c]: No worries

[the remark this answers is not in the document]
Comment (i10416) [reply clib/src/main/resources/scala-native/fenv.c]: Thank you for reviewing :)

**Review by WojciechMazur — APPROVED** [review]

**Comment by LeeTibbert** [comment]
@i10416 

I realize that this PR merged almost a year ago.  First, as a member of the SN community, thank you for the PR.

Once upon a time, Scala Native and its tests ran on the FreeBSD operating system.  FreeBSD was never part
of the Continuous Integration (CI) matrix, so failures to run on FreeBSD crept in over time as code was added.


I am trying to get the SN unit-tests running cleanly on FreeBSD again and have a question.
Thought I would contact you, the original author  to ask your help/thoughts.

1) The file is called `FenvTest.scala` but the class is called `FEnvTest` (note the 'E', not 'e').  I noticed this
     when trying to run the class based on the file name.  May I change one name to the other, so both
     match.  Should I change the class to match the file name?  

Thank you for your recommendation.

I would normally not bother the author of a long since merged PR for such a small issue.
I am in that file trying to figure out if a particular feMumble symbol is missing on both FreeBSD X86_64
and arm or only arm architecture.

**Pull request merged**